packages/fuels-programs/src/lib.rs

```diff
@@ -21,3 +32,323 @@ impl Configurables {
+            Ok(decoded_value)
+        } else {
+            Err(Self::parsing_error(output_param))
+        }
+    }
+
+    fn parsing_error(output_param: &ParamType) -> Error {
+        error!(
+            InvalidData,
+            "ReceiptDecoder: failed to find matching receipts entry for {:?}", output_param
+        )
+    }
+
+    fn extract_return_data(&mut self, contract_id: ContractId) -> Option<Vec<u8>> {
+        let (index, data) = self.receipts.iter().enumerate().find_map(|(i, receipt)| {
+            if matches!(receipt,
+                    Receipt::ReturnData { id, .. } if *id == contract_id)
+            {
+                Some((
+                    i,
+                    receipt
+                        .data()
+                        .expect("ReturnData should have data")
+                        .to_vec(),
+                ))
+            } else {
+                None
+            }
+        })?;
+
+        self.receipts.remove(index);
+        Some(data)
+    }
+
+    fn extract_return(&mut self, contract_id: ContractId) -> Option<Vec<u8>> {
+        let (index, data) = self.receipts.iter().enumerate().find_map(|(i, receipt)| {
+            if matches!(receipt,
+                    Receipt::Return { id, ..} if *id == contract_id)
+            {
+                Some((
+                    i,
+                    receipt
+                        .val()
+                        .expect("Return should have val")
+                        .to_be_bytes()
+                        .to_vec(),
+                ))
+            } else {
+                None
+            }
+        })?;
+
+        self.receipts.remove(index);
+        Some(data)
+    }
+
+    fn extract_return_data_heap(&mut self, contract_id: ContractId) -> Option<Vec<u8>> {
+        // If the output of the function is a vector, then there are 2 consecutive ReturnData
+        // receipts. The first one is the one that returns the pointer to the vec struct in the
+        // VM memory, the second one contains the actual vector bytes (that the previous receipt
+        // points to).
+        // We ensure to take the right "first" ReturnData receipt by checking for the
+        // contract_id. There are no receipts in between the two ReturnData receipts because of
+        // the way the scripts are built (the calling script adds a RETD just after the CALL
+        // opcode, see `get_single_call_instructions`).
+        if let Some((i, vector_data)) = self.receipts.iter().enumerate().tuple_windows().find_map(
+            |((i, current_receipt), (_, next_receipt))| {
+                let vec_data = Self::extract_vec_data(current_receipt, next_receipt, contract_id)?;
+                Some((i, vec_data.clone()))
+            },
+        ) {
+            self.receipts.remove(i);
+            // the index has adjusted due to the previous removal
+            self.receipts.remove(i);
+            Some(vector_data)
+        } else {
+            None
+        }
```

Comment:
```suggestion
        for (index, (current_receipt, next_receipt)) in
            self.receipts.iter().tuple_windows().enumerate()
        {
            if let Some(data) = Self::extract_vec_data(current_receipt, next_receipt, contract_id) {
                let data = data.clone();
                self.receipts.drain(index..=index + 1);
                return Some(data);
            }
        }

        None
```
Notes about the suggestion:
* Removes both elements at once  via `drain`, no need for the comment on why `i` remains the same
* seems simpler to do windows first and then enumerate them instead of the other way around
* for loop seems to cut down on the noise, allows for removing from vectors and immediately returning without worrying about the state of the iterator.